## Homework

Using the Newton-Raphson method, find the solution for  
\begin{eqnarray}
x_{1}^2+x_{1}x_{2}=10\\
x_{2}+3x_{1}x_{2}^2=57
\end{eqnarray}

Hint: first you need to derive the partical derivative of the equation arrays. You can do it either 
analytically by hand or numerically using method we learned in the last lecture.

In [22]:
import numpy as np

# The first function
def f(x1, x2):
    return x1*x1 + x1*x2 - 10

# The second function
def g(x1, x2):
    return x2 + 3*x1*x2*x2 - 57

# The derivative of f with respect to x1
def dfdx1(x1, x2):
    return 2*x1 + x2

# The derivative of f with respect to x2
def dfdx2(x1, x2):
    return x1

# The derivative of g with respect to x1
def dgdx1(x1, x2):
    return 3*x2*x2

# The derivative of g with respect to x2
def dgdx2(x1, x2):
    return 1 + 6*x1*x2

# Returns a vector of the functions' outputs.
def Function(x):
    return np.array([f(x[0], x[1]), g(x[0], x[1])])

# Returns the jacobian of the two functions
def Jacobian(x):
    return np.array([[dfdx1(x[0], x[1]),dfdx2(x[0], x[1])],[dgdx1(x[0],x[1]), dgdx2(x[0],x[1])]])

#===========================================================
# Parameters:
# - x: An initial guess for a vector that satisfies the
#      equations
# - F: A vector form of the input equations 
# - J: The jacobian of F
# - eps: The desired tolerance for error
# - Nmax: The maximum number of times to iterate the Newton
#         -Raphson method.
# Returns:
#   A vector of solutions that satisfy the given equations
#===========================================================
def dim_2_newton_raphson(x, Function, Jacobian, eps, Nmax):
    
    iter_num = 0
    while np.linalg.norm(Function(x)) > eps and iter_num < Nmax:
        f = Function(x)
        j = Jacobian(x)
        jacobian_determinant = (j[0,0]*j[1,1] - j[0,1]*j[1,0])
        dx1 = (j[0,1]*f[1] - j[1,1]*f[0]) / jacobian_determinant
        dx2 = (j[1,0]*f[0] - j[0,0]*f[1]) / jacobian_determinant
        x += np.array([dx1, dx2])
        iter_num += 1
    
    if iter_num >= Nmax:
        print("Did not converge after {} iterations".format(Nmax))
    
    return x   


xi = np.array([1, 1.])
x1, x2 = dim_2_newton_raphson(xi, Function, Jacobian, 1e-8, 100)
print(x1, x2)
print(f(x1,x2), g(x1,x2))
print()
xi = np.array([-1, -1.])
x1, x2 = dim_2_newton_raphson(xi, Function, Jacobian, 1e-8, 100)
print(x1, x2)
print(f(x1,x2), g(x1,x2))
print()

2.000000000000006 2.9999999999999907
2.4868995751603507e-14 -1.7763568394002505e-13

4.393744193289869 -2.1177810147165195
-1.7905676941154525e-12 1.4516388091578847e-10



# Results

Running through a few different starting locations I found the roots (2,3) and [approximately] (4.394,-2.118). Because these equations are quadratic in both variables, they should have exactly two solutions.